File: 240229/G or H 2/G-or-H-2.cpp
#include <iostream>

using namespace std;

char alphabetArray[101];

bool isInRange(int start, int distance){
    return 0 <= start && start <= 100 
    && 0 <= start+distance && start+distance <= 100;
}

bool checkIsMatched(int start, int distance){
    int GCount = 0;
    int HCount = 0;
    for(int i = start; i <= start+distance; i++){
        if(alphabetArray[i] == 'G'){
            GCount++;
        }
        if(alphabetArray[i] == 'H'){
            HCount++;
        }
    }
    if(GCount * HCount == 0){
        return true;
    }
    if(GCount == HCount){
        return true;
    }
    return false;
}

int main() {
    int N;
    cin >> N;
    while(N--){
        int position;
        char alphabet;
        cin >> position >> alphabet;
        alphabetArray[position] = alphabet;
    }

    int maxDist = 0;
    for(int i = 0; i <= 100; i++){
        for(int distance = 0; distance <= 100; distance++){
            if(!isInRange(i, distance)){
                continue;
            }
            if(alphabetArray[i] == '\0' || alphabetArray[i+distance] == '\0'){
                continue;
            }
            
            bool isMatched = checkIsMatched(i, distance);
            if(isMatched){
                maxDist = max(maxDist, distance);
            }
        }
    }

    cout << maxDist;
    return 0;
}

File: 250505/G or H 2/G-or-H-2.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        int N;
        char[] personAlphabets;

        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        personAlphabets = new char[101];

        for (int i = 0; i < N; i++) {
            int personPos = sc.nextInt();
            char alphabet = sc.next().charAt(0);
            personAlphabets[personPos] = alphabet;
        }
        sc.close();

        new Solver(personAlphabets).solve();
    }
}

class Solver {
    int personAlphabetsIndex;
    char[] personAlphabets;

    public Solver(
            char[] personAlphabets
    ) {
        this.personAlphabets = personAlphabets;
        this.personAlphabetsIndex = personAlphabets.length - 1;
    }

    public void solve() {
        int maxDist = 0;
        for (int start = 0; start <= personAlphabetsIndex; start++) {
            for (int end = start; end <= personAlphabetsIndex; end++) {
                if (isPersonNotExistAtBorder(start, end)) {
                    continue;
                }
                if (isValid(start, end)) {
                    maxDist = Math.max(maxDist, end - start);
                }
            }
        }
        System.out.println(maxDist);
    }

    private boolean isPersonNotExistAtBorder(int start, int end) {
        return personAlphabets[start] == 0 || personAlphabets[end] == 0;
    }

    private boolean isValid(int start, int end) {
        int gCount = 0;
        int hCount = 0;

        for (int i = start; i <= end; i++) {
            if (personAlphabets[i] == 'G') {
                gCount++;
            }
            if (personAlphabets[i] == 'H') {
                hCount++;
            }
        }

        if (gCount == 0 && 0 < hCount) {
            return true;
        }
        if (hCount == 0 && 0 < gCount) {
            return true;
        }
        if (hCount == gCount) {
            return true;
        }
        return false;
    }
}